students/eowyn/session10/ff-mailroom/ffmailroom.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python
+import sys
+import copy
+from textwrap import dedent
+
+
+class Donor():
+    # Keep track of individual donors and their history of donations
+
+    def __init__(self, name):
+        self.donations = []
+        self.name = ''.join(name)  # should we check input?
+
+    def add_donation(self, amt):
+        self.donations.append(float(amt))
+
+    def generate_letter(self):
+        '''
+        Generate a formatted thank you note to a donor and return
+        the string of the thank you note for printing or writing.
+        '''
+        fs = "Thank you, {0}, for your generosity and recent gift of ${1:.2f}."
+        return fs.format(self.name, self.most_recent_donation)
+
+    def mult_donations(self, factor):
+        self.donations = list(map(lambda x: x * factor, self.donations))
+
+    def filter_donations(self, min_donation, max_donation):
+        if min_donation > max_donation:
+            (min_donation, max_donation) = (max_donation, min_donation)
+        self.donations = list(filter(lambda x: x > min_donation and
+                                     x < max_donation, self.donations))
+
+    @property
+    def sum_donations(self):
+        return round(sum(self.donations), 2)
+
+    @property
+    def count_donations(self):
+        return len(self.donations)
+
+    @property
+    def most_recent_donation(self):
+        return round(self.donations[-1], 2)
+
+    @property
+    def average_donation(self):
+        return round(self.sum_donations / self.count_donations, 2)
+
+
+class Transactions():
+    # Collect list of Donor objects and generate reports
+
+    def __init__(self):
+        self.all_donors = []
+
+    def add_donor(self, name, amt):
+        this_donor = self.get_donor(name)
+        if this_donor:
+            this_donor.add_donation(amt)
+        else:
+            this_donor = Donor(name)
+            this_donor.add_donation(amt)
+            self.all_donors.append(this_donor)
+        return this_donor.generate_letter()
+
+    def get_donor(self, name):
+        for i in self.all_donors:
```

Comment:
maybe a more meaningful name than "i" could be good -- not a big deal with this little code, but still more clear.